compair/tests/api/test_impersonation.py

```diff
@@ -0,0 +1,204 @@
+        self.assert200(rv)
+        self.assertEquals(rv.json.get('impersonating'), False)
+
+    def test_impersonation_requires_login(self):
+        self.assert401(self._start_impersonation(self.instructor_main_course))
+        self.assert401(self._start_impersonation(self.student_main_course))
+
+    def test_student_cannot_impersonate(self):
+        with self.login(self.student_main_course.username):
+            self.assert403(self._start_impersonation(self.instructor_main_course))
+            self.assert403(self._start_impersonation(self.student_second_course))
+
+    def test_instructor_can_impersonate_own_students(self):
+        with self.login(self.instructor_main_course.username):
+            self.assert200(self._start_impersonation(self.student_main_course))
+            self._verify_impersonation(original=self.instructor_main_course, pretending=self.student_main_course)
+
+    def test_instructor_cannot_impersonate_student_not_in_his_course(self):
+        with self.login(self.instructor_main_course.username):
+            self.assert403(self._start_impersonation(self.student_second_course))
+        with self.login(self.instructor_second_course.username):
+            self.assert403(self._start_impersonation(self.student_main_course))
+
+    def test_instructor_cannot_impersonate_admin(self):
+        with self.login(self.instructor_main_course.username):
+            self.assert403(self._start_impersonation(DefaultFixture.ROOT_USER))
+
+    def test_admin_can_impersonate_students_but_not_instructors(self):
+        with self.login(DefaultFixture.ROOT_USER.username):
+            self.assert200(self._start_impersonation(self.student_main_course))
+            self.assert403(self._start_impersonation(self.instructor_main_course))
+
+    def test_end_impersonation(self):
+        # Impersonation will block non-GET traffic, but should allow end impersonation (a DELETE call)
+        with self.login(self.instructor_main_course.username):
+            self.assert200(self._start_impersonation(self.student_main_course))
+            self._verify_impersonation(original=self.instructor_main_course, pretending=self.student_main_course)
+
+            self.assert200(self._end_impersonation())
+            self._is_not_impersonating()
+
+    def test_cannot_logout_during_impersonation(self):
+        with self.login(self.instructor_main_course.username):
+            self.assert200(self._start_impersonation(self.student_main_course))
+            self._verify_impersonation(original=self.instructor_main_course, pretending=self.student_main_course)
+
+            # explicit logout
+            self.assert403(self.logout())
+            self._verify_impersonation(original=self.instructor_main_course, pretending=self.student_main_course)
+            self.assert200(self._end_impersonation())
+            self.assert200(self.logout())
+
+    def test_impersonation_access_assignment(self):
```

Comment:
These access tests shouldn't be here. Instead, every api endpoint test needs to be updated to include impersonation checking for an authorized instructor impersonating an authorized student (to ensure that POST, DELETE and GET requests are handled properly)

In `ComPAIRAPITestCase` you need to add a new context manager
```
    @contextmanager
    def impersonate(self, original_user, target_user):
        self.login(original_user.username):
            rv = self.client.post('/api/impersonate/'+target_user.uuid,
                content_type='application/json', follow_redirects=True)
            self.assert200(rv)
            yield rv
            self.client.delete('/api/impersonate', follow_redirects=True)
```

so that you can do 
```
        with self.impersonate(self.data.get_authorized_instructor(), self.data.get_authorized_student()):
            #rv = self.client.post(self.url, data=json.dumps(...), content_type='application/json')
            #self.assert403(rv)
            #rv = self.client.delete(self.url, content_type='application/json')
            #self.assert403(rv)
            #rv = self.client.get(self.url, content_type='application/json')
            #self.assert200(rv) #or self.assert403(rv) if normally not accessible to students
```